### ISSUE TITLE
REPL hangs with interference with another socket / event loop

### ISSUE THREAD

**Issue by zwang** [omcljs/ambly#106]
Hi Mike, 

Thank you for the cool library. I am trying to use this in my project. However, after I made a choice, the repl hangs there (seems forever). After I terminate the app, the connection reset socket exception is thrown. The ambly volume (Ambly-B65AF28D) is successfully mounted though and I opened it I can see the files from the folder I specified.  Here is what looks like in the console (No exception either):

```
To quit, type: :cljs/quit
[1] Pla on iPhone 5s
[R] Refresh
Choice: 1
Connecting to Pla on iPhone 5s ...
```

I am using OS X El Capitan 10.11.1. I start the repl using `boot ambly`. My iphone is 5s with iOS 8.4.1. My app is a swift based app using Swift 2.1 with XCode 7.1.1. The same problem happens on iOS Simulator 5s 8.1

**Comment by mfikes** [comment]
Hi @zwang, can you check https://github.com/omcljs/ambly/wiki/Connectivity to see if any of it might apply to your environment?

**Comment by zwang** [comment]
Yeah, I checked before I posting the question.I didn't see anything matching my case. I just double checked again and ran the `dns-sd` command, got the following result:

```
➜  Clojure git:(master) dns-sd -B _http._tcp
Browsing for _http._tcp
DATE: ---Wed 11 Nov 2015---
9:11:21.825  ...STARTING...
Timestamp     A/R    Flags  if Domain               Service Type         Instance Name
9:11:21.826  Add        3   4 local.               _http._tcp.          D-Link DIR-860L Configuration Utility
9:11:21.826  Add        3   4 local.               _http._tcp.          D-Link SharePort Web Access
9:11:21.826  Add        2   4 local.               _http._tcp.          Ambly Pla on iPhone 5s
9:15:34.492  Rmv        0   4 local.               _http._tcp.          Ambly Pla on iPhone 5s
9:16:16.434  Add        2   4 local.               _http._tcp.          Ambly Ambly Demo on iPhone 5s 
```

I also tried to connect to Ambly Demo app on iPhone 5s. It actually connects. The code for setting up the server is shown below. I am using a JSContext to get JSGlobalContextRef. Also I am using GCDWebserver 3.2.5 other than 3.2.2 in the demo project since my project is using Carthage (github "swisspol/GCDWebServer" ~> 3.2.5) :

```
func setupABYServerForJSContext(context: JSContext, outputDir: NSURL!) {
    #if DEBUG
        abyContextManager = ABYContextManager(context: context.JSGlobalContextRef, compilerOutputDirectory: outputDir)
        abyContextManager!.setupGlobalContext()
        abyContextManager!.setUpConsoleLog()
        abyContextManager!.setUpTimerFunctionality()
        abyContextManager!.setUpAmblyImportScript()
        abyServer = ABYServer(context: abyContextManager!.context, compilerOutputDirectory: outputDir)
        let isSuccessful = abyServer!.startListening()
        if !isSuccessful {
            Logger.info("Failed to start REPL Server")
        }
    #endif
}
```

One thing I noticed is in the demo app, when I connected, there are much more warnings in the Xcode console:

```
[WARNING] Unknown DAV property requested "quota-available-bytes"
[WARNING] Unknown DAV property requested "quota-used-bytes"
[WARNING] Unknown DAV property requested "quota"
[WARNING] Unknown DAV property requested "quotaused"
[WARNING] Unknown DAV property requested "quota-available-bytes"
[WARNING] Unknown DAV property requested "quota-used-bytes"
[WARNING] Unknown DAV property requested "quota"
[WARNING] Unknown DAV property requested "quotaused"
[WARNING] Unknown DAV property requested "getetag"
[WARNING] Unknown DAV property requested "getetag"
[WARNING] Unknown DAV property requested "getetag"
[WARNING] Unknown DAV property requested "getetag"
[WARNING] Unknown DAV property requested "getetag"
[WARNING] Unknown DAV property requested "getetag"
[WARNING] Unknown DAV property requested "getetag"
[WARNING] Unknown DAV property requested "getetag"
[WARNING] Unknown DAV property requested "getetag"
[WARNING] Unknown DAV property requested "getetag"
[WARNING] Unknown DAV property requested "getetag"
[WARNING] Unknown DAV property requested "getetag"
[WARNING] Unknown DAV property requested "getetag"
[WARNING] Unknown DAV property requested "getetag"
[WARNING] Unknown DAV property requested "getetag"
[WARNING] Unknown DAV property requested "getetag"
[WARNING] Unknown DAV property requested "getetag"
```

However, in my app, there are only 4 warnings:

```
[WARNING] Unknown DAV property requested "quota-available-bytes"
[WARNING] Unknown DAV property requested "quota-used-bytes"
[WARNING] Unknown DAV property requested "quota"
[WARNING] Unknown DAV property requested "quotaused"
```

Thank you for your help.

**Comment by mfikes** [comment]
@zwang Perhaps it is connecting to your simulator, but getting stuck in some way when bootstrapping the ClojureScript environment within it. I'd suggest deleting the app from your simulator (or better yet, Reset Content and Settings) and trying again with the simulator.

If you look at the GCDWebServer repo, you'll see many issues mentioning Swift that cropped up; perhaps you are seeing Swift issues with Ambly's code itself... dunno.

**Comment by zwang** [comment]
@mfikes  I will try your advice soon and report back. 

I also just created a demo project using Swift 2.1 and Xcode 7.1.1 using CocoaPods and copying the same code from my project. It actually works like a charm. However, my project still not working. I am wondering if it is because my app also using TCP socket communicating server. I will think about how to verify this assumption. If you have any advice for this, please let me know. 

Thank you very much for your help. Really appreciate it. :)

**Comment by mfikes** [comment]
@zwang Ahh. Good point. I'm not an expert on this aspect, but perhaps run loops are interfering? For reference, here is how Ambly sets its up:

https://github.com/omcljs/ambly/blob/master/ObjectiveC/src/ABYServer.m#L426-L444

Maybe try disabling the TCP aspect of your app temporarily to see...

**Comment by zwang** [comment]
@mfikes Yeah, it is because of the socket in my app. When I start the REPL server at the beginning of app starts, I can connect to it successfully and send cljs commands over and get result back. And after my app socket connected, I can still eval code in REPL and get results back. 

So there is definitely some interfering going on there. I will look into it this issue. Our socket handling function are GCD based with `DISPATCH_IO_STREAM`, `dispatch_io_read` , `dispatch_io_write`. The connecting is done using C api like below. And so far we only do IPv4.  I am not good on this aspect  at all. If you have any advice, I am all ears. :)

```
static func socketer(error: NSErrorPointer = nil) -> CInt? {
    let s: CInt = socket(PF_INET, SOCK_STREAM, IPPROTO_TCP)  // IPv4
    if (s == -1) {
        if error != nil {
            error.memory = lastErr("socket(...) failed.")
        }
        return nil
    }
    setsock(s, option: SO_NOSIGPIPE, flag: true)
    setnonblock(s)
    Logger.info("socketer: socket \(s) initialized")
    return s
}

static func connecter(socket: CInt, endpoint: Endpoint) -> CInt {
    var addr = sockaddr_in(
    sin_len: __uint8_t(sizeof(sockaddr_in)),
            sin_family: sa_family_t(AF_INET),
            sin_port: port_htons(endpoint.port),
            sin_addr: in_addr(s_addr: endpoint.address.toNet4()),
            sin_zero: (0, 0, 0, 0, 0, 0, 0, 0)
    )

    var sock_addr = sockaddr(sa_len: 0, sa_family: 0, sa_data: (0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0))
    memcpy(&sock_addr, &addr, Int(sizeof(sockaddr_in)))
    return connect(socket, &sock_addr, socklen_t(sizeof(sockaddr_in)))
}
```

I have to start the REPL server after my app connected via socket to the server since I create and destroy JSContext adhoc. 

Thanks a lot for the help. You are awesome. Same as this library. :+1:

**Comment by mfikes** [comment]
@zwang I don't have any iOS server socket experience apart from the stuff you see in Ambly. Interestingly, Ambly's server socket appears to be fine with the WebDAV server socket employed by the GCDWebServer used by Ambly. Hrm. Maybe if you can provide a minimal repro (maybe a little additional socket code in `AppDelegate.m` for the Ambly Demo app or some such) that can cause the lockup, that would help.

**Comment by zwang** [comment]
@mfikes Yes, I will try to provide a minimal repo later today. Thanks for all the help. :)

**Comment by zwang** [comment]
Hi @mfikes 

Some updates: I figured out what is the issue. It is because I am not starting the server in the main thread. Once I started this in the main thread, it works. I build a minimal demo app just like our main app with the same client network and server and it worked well with the ambly server last night. So I started thinking what are the difference between the two. And I recalled that background thread can not be used to connect network in iOS. then I did some testing, it affirmed my guess. Hope my feedback can be helpful to somebody.

Now I have a reason to use clojurescript at work with swift and ios, great! 

Thank you again for your help and all the great work. I have been reading your blogs and watching your videos. getting exciting about the movements of cljs and react native.
